### PR TITLE
Provide Pipeline metadata controller to update metadata from Jenkins

### PR DESCRIPTION
**Pull request by JohnNiang**
### What this PR dose / why we need it?

Provide Pipeline metadata controller to update metadata from Jenkins in every 10 seconds.

And this controller only cares about the newly created Pipelines and the Pipelines that can be obtained when it is restarted.

### Steps to test

1. Replace devops controller with docker image `johnniang/devops-controller:dev-v3.2.0-alpha.0-ca751a7`
2. See annotations of Pipeline resources

### TODO

- [ ] Bump version of Jenkins client after https://github.com/jenkins-zh/jenkins-client/pull/27 gets merged

/kind feature

### CODE REVIEW

**Review by LinuxSuRen — CHANGES_REQUESTED** [review]
It works. But there are some fields that might be useless for now. I'd suggest we remove them. Such as:

* permissions
* DisplayName
* fullName
* Organizatioin

```
     10     pipeline.devops.kubesphere.io/jenkins-metadata: '{"organization":"jenkins","name":"simple-2","displayName":"simple-2","fullName":"testpzwjf/simple-2","full        DisplayName":"testpzwjf/simple-2","weatherScore":100,"estimatedDurationInMillis":-1,"permissions":{"configure":true,"create":true,"read":true,"start":true,"sto        p":true}}
```

## Why we should remove them

Useless data can let the etcd has high pressure.

**Comment by JohnNiang** [comment]
I've exclude some fields of Pipeline's metadata according to https://github.com/kubesphere/ks-devops/pull/282#pullrequestreview-766151872, please see the result:

```json
{
   "estimatedDurationInMillis":8282,
   "name":"github-pipeline",
   "numberOfPipelines":4,
   "totalNumberOfBranches":4,
   "numberOfFailingBranches":1,
   "totalNumberOfPullRequests":1,
   "branchNames":[
      "master",
      "PR-1",
      "sonarqube",
      "podman",
      "v0.0.1"
   ],
   "scmSource":{
      "id":"github",
      "apiUrl":"https://api.github.com"
   },
   "scriptPath":"Jenkinsfile-online"
}
```

```json
{
   "weatherScore":100,
   "estimatedDurationInMillis":3839,
   "name":"gitlab-pipeline",
   "numberOfPipelines":1,
   "totalNumberOfBranches":1,
   "totalNumberOfPullRequests":2,
   "branchNames":[
      "main",
      "MR-2",
      "MR-3"
   ],
   "scmSource":{
      
   },
   "scriptPath":"Jenkinsfile"
}
```

File: controllers/jenkins/pipeline/pipeline_metadata_controller.go [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+		if err := r.Get(context.Background(), pipelineKey, pipeline); err != nil {
+			return client.IgnoreNotFound(err)
+		}
+		metadataJSON, err := json.Marshal(metadata)
+		if err != nil {
+			return err
+		}
+		// diff pipeline metadata
+		if pipeline.Annotations[v1alpha3.PipelineJenkinsMetadataAnnoKey] == string(metadataJSON) {
+			// skip updation if metadata unchanged
```
Comment (LinuxSuRen) [inline controllers/jenkins/pipeline/pipeline_metadata_controller.go:86]: typo `updation`

> skip updation if metadata unchanged
 
`skip update if the metadata was unchanged`

File: controllers/jenkins/pipeline/pipeline_metadata_controller.go [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+	metadata := convert(jobPipeline)
+	return retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		pipeline := &v1alpha3.Pipeline{}
+		if err := r.Get(context.Background(), pipelineKey, pipeline); err != nil {
+			return client.IgnoreNotFound(err)
+		}
+		metadataJSON, err := json.Marshal(metadata)
+		if err != nil {
+			return err
+		}
```
Comment (LinuxSuRen) [inline controllers/jenkins/pipeline/pipeline_metadata_controller.go:83]: I prefer to move these lines be outside of the `RetryOnConflict`.

File: controllers/jenkins/pipeline/pipeline_metadata_controller.go [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+	}
+	// fetch pipeline metadata from Jenkins
+	boClient := job.BlueOceanClient{
+		JenkinsCore:  r.JenkinsCore,
+		Organization: "jenkins",
+	}
+	pipelineMetadata, err := boClient.GetPipeline(pipeline.Name, pipeline.Namespace)
+	if err != nil {
+		log.Error(err, "unable to get Pipeline metadata from Jenkins")
+		r.recorder.Eventf(pipeline, v1.EventTypeWarning, FailedMetaUpdate, "Failed to update metadata of Pipeline from Jenkins")
```
Comment (LinuxSuRen) [inline controllers/jenkins/pipeline/pipeline_metadata_controller.go:57]: Do we really need a similar log in two places?

**Review by LinuxSuRen — COMMENTED** [review]
I left some small suggestions. I don't block this PR. But please feel free to take a look at it.

File: controllers/jenkins/pipeline/pipeline_metadata_controller.go (hunk 3 above)
Comment (JohnNiang) [reply controllers/jenkins/pipeline/pipeline_metadata_controller.go:57]: My bad. I've forgotten the err.

File: controllers/jenkins/pipeline/pipeline_metadata_controller.go (hunk 2 above)
Comment (JohnNiang) [reply controllers/jenkins/pipeline/pipeline_metadata_controller.go]: That's great.

**Comment by LinuxSuRen** [comment]
The image for test purposes:
```
surenpi/devops-controller:dev-v3.2.0-alpha.1-9fa4a0b
```

**Review by LinuxSuRen — COMMENTED** [review]
/lgtm

**Pull request merged**